Sprint 23(home work) Gr26

amanzhola opened the pull request.

amanzhola committed:
Sprint 22(home work) Gr26
🙏 ➡️ 🔥 💃 (👀💎!✨)

1. Исправлена логика перехода FragmentPlaylist на CreatePlaylistFragment и обратно FragmentPlaylist ->
Убрана восстановление скролла через saveEnabled="false" лишает гибкости при высокой эффективности работы ресурса
Самое трудное было найти идею: сделать так чтобы при переключении темы сохранялось позиция, а при входе извне было на верху списка включая переключение извне. В итоге при переходе в CreatePlaylistFragment на FragmentPlaylist убираем из стак старый MediaLibraryFragment
2. Подхват у родителя MediaLibraryFragment который имеет тулбар от BaseActivity
3. Реализация в FragmentPlaylist перехода из CreatePlaylistFragment

Благодаря этому теперь можно отличить переключение темы "извне" [нужно ставить на верх списка] от "внутри" [сохранить позицию какая была включая поворот экрана как и переключение темы]

4. очищен fragment_playlist.xml от принудительного saveEnabled="false"
5. Попытки восстановить тулбар для голого FragmentPlaylist вырванного из стак(первая попытка, далее стало понятно что нужно весь MediaLibraryFragment убрать) через BaseActivity и BaseFragment - бесполезная трата времени (30% от всего спринта)
6. Constants.kt - добавление ключа для подхвата у родителя табов MediaLibraryFragment
7. strings.xml (просто добавление emoji для украшения будущих строк)

amanzhola committed:
Sprint 22(home work) Gr26

1. Добавлено база данных избранных, Bottom Sheet c навигацией на CreatePlaylistFragment и обратно, база данных альбомов для TrackPreviewActivity (просмотр треков песен и музыки полученых по обмену)
-> Было бы легко провести на ExtraOptionFragment (аудио) где все готово и остается скрыть поле ввода и сменить имя, но это территория ревьюеров поэтому сделана отдельная независимая версия

2. Исправлена логика CreatePlaylistFragment так как жесткая привязка на возврат к FragmentPlaylist нарушала ТЗ по возврату ExtraOptionFragment (аудио) + добавлен для просмотра музыки и песен полученных через обмен TrackPreviewActivity

3. Исправлен баг когда при переходе из FragmentFavouriteTracks на ExtraOptionFragment(просмотр треков) и далее возврате терялся цвет заглавия тулбара у родителя таба MediaLibraryFragment

Updates done:
1. AppModule
TrackPreviewViewModel + 3 more DI for DB(ObservePlaylistsUseCase,
AddTrackToPlaylistUseCase) and for favourite(FavoriteTracksInteractor)
2. TrackPreviewViewModel
fixed play button (PREPARING & PREPARED = isPlaying = true)
+ ❤️ Логика для кнопки "лайк"
+ 📜 👉 📝 ПЛЕЙЛИСТЫ (НОВОЕ)
3. ExtraOptionViewModel
-> cleaning by taking off onFavoriteClicked() without parameters
4. MainActivity
+ 3 steps to reach from TrackPreviewActivity via MainActivity to CreatePlaylistFragment
handleExternalIntent(intent)
5. TrackAdapter
"from_favorites" to true // 🟢 Новый флаг! -> useless attemp to fix MediaLibraryFragment title
"IS_FROM_SEARCH" to true, // 🟢 Новый флаг!-> add to pas flag to CreatePlaylistFragment
6. ExtraOptionFragment
removed viewModel.onFavoriteClicked() -> excess
created requireActivity().onBackPressedDispatcher.onBackPressed() try
to handle title color issue FragmentFavouriteTracks(wrong attemps, problems was color visibity)
7. TracksDiffCallbackAudio
created additional if (oldItem.isFavorite != newItem.isFavorite) -> removed
8. ExtraOptionFragment
to get "from_extra" via checking back stack
From CreatePlaylistFragment to be return back to ExtraOptionFragment
9. CreatePlaylistFragment
fix the navigation from CreatePlaylistFragment to FragmentPlaylist
fix the navigation from CreatePlaylistFragment to TrackPreviewActivity
fix the navigation from CreatePlaylistFragment to ExtraOptionFragment
fix the navigation to the rest
10. TrackAdapterTD
attemp to get from LauncherActivity to TrackPreviewFragment (transfer of TrackPreviewActivity)
for getting to CreatePlaylistFragment using 2 navigation: nav_track_detail.xml
and playlist_navigation_graph.xml (wrong step -> removed)
11. TrackDetailActivity
-> transfer to TrackPreviewFragment (deleted -> wrong step)
-> Bottom Sheet + (removed)
-> attemps to get TrackPreviewFragment via MainActivity to CreatePlaylistFragment and back via TrackPreviewActivity
all updated removed (since wrong)
12. TrackPreviewActivity (share audio tracks preview)
-> done Bottom sheet
-> done onFavoriteClicked(track: Track)
-> done onAddTrackClicked(track: Track)
-> done navigation from TrackPreviewActivity to CreatePlaylistFragment via MainActivity(Single Activity)
-> Snackbar
and back via registerForActivityResult (createLauncher)
13. TrackPreviewFragment
transfer from TrackPreviewActivity (removed and deleted)
14. MediaLibraryFragment
-> added to resume applyToolbarTheme() to fix title color issue on return to FragmentFavouriteTracks from ExtraOptionFragment
15. MediaLibraryPagerAdapter
cleaning from prev updates
16. FragmentFavouriteTracks
cleaning from prev updates
17. FragmentPlaylist
-> moving from CreatePlaylistFragment and back to FragmentPlaylist
-> saving toolbar by removing from stack MediaLibraryFragment with child FragmentPlaylist to update
-> to have night mode on top of list and same time on night mode inside the FragmentPlaylist on it's position
logic task: to be on it's position for screen turn and night mode and same time be on top in case enter from outside
18. object NavDebug
for logging - removed
19. activity_track_preview.xml + bottom sheet
20. fragment_track_preview.xml + bottom sheet (removed)
21. playlist_navigation_graph.xml + trackPreviewFragment (removed)
22. strings.xml + track_preview (removed -> excess)
23. fragment_extra_option.xml -> clean from prev updates

amanzhola committed:
Sprint 22(home work) Gr26

1. MediaLibraryFragment: возвращена возврат через заглавие в MainActivity (сетку кнопок) в резюме добавлен пропущенная конфигурция возврата в тулбаре
2. Добавлена закоментированная опция эмодзи-цветок для выпадающей кнопки выбора в тулбаре без опции селектора (постоянная видимость) в menu.xml via title + string.xml
3. Подчищены комменты в TrackPreviewViewModel (цвет заглавия, разделительной линии остальной логики), в MediaLibraryFragment (resume -> title and bg colors)

amanzhola committed:
Sprint 23(home work) Gr26

1. AndroidManifest.xml 3 варианта ЛОГО
2. base_main.xml (ничего)
3. bg_dialog_square.xml (квадрат для диалога и управление его цветом)
4. Constants.kt -> for const
5. CreatePlaylistFragment -> ТЗ вариант 2 редактирование через логику
6. CreatePlaylistViewModel см выше
7. dimens.xml -> соотвествие: маленькая, сред и таб
8. dot_3.xml -> 3 dots for menu on PlaylistInfoFragment
9. ExtraOptionFragment -> переходы на CreatePlaylistFragment
10. fragment_create_playlist.xml (ничего на 23)
11. fragment_playlist_info.xml -> общий и 1 для табов. Просмор информации альбомов
12. FragmentPlaylist -> переходs для ТЗ 23
13. heart.xml, icon_launcher.xml, heart_launcher.xml -> 3 опции ЛОГО
14. integers.xml -> управление кол-во строк и также кол-во сеток для FragmentPlaylist
15. item_menu_action.xml -> строки для 2ой BS PlaylistInfoFragment
16. item_playlist_bottom.xml -> cards for BS
17. launcher_bg.xml -> white color + textcolor on Theme
18. MainActivity -> hiding bottom sheet navig for PlaylistInfoFragment
19. MediaStoreSaver_public (public path for picture memory saver)
20. MenuAdapter -> for 2nd BS in PlaylistInfoFragment to recycle card and 3 texts for options edit, delete and share
21. MenuRow for MenuAdapter
22. playlist_navigation_graph.xml + PlaylistInfoFragment + navigation paths
23. PlaylistInfoViewModel.kt new
24. PlaylistMappers.kt update
25. PlaylistRepositoryImpl update
26. playlistsModule + new PlaylistInfoViewModel
27. PlaylistTrackDao + update
28. strings.xml + update
29. TrackAdapter + long click on delete
30. TrackPreviewActivity + exchange with songs and movies
31. UpdatePlaylistUseCase for edit on CreatePlaylistViewModel

anastasianadelshtehel commented on app/src/main/java/com/example/playlistmaker/presentation/playlistInfo/PlaylistInfoViewModel.kt:
```diff
@@ -0,0 +1,65 @@
+        val description: String?,
+        val coverPath: String?,
+        val minutesTotal: Long,   // минуты (сырые)
+        val tracksCount: Int,     // количество треков
```
Поле tracksCount дублирует информацию, которую можно получить из tracks.size. эТо может привести к расхождению данных. Лучше убрать tracksCount и добавить вычисляемое свойство 
    **val tracksCount: Int get() = tracks.size**

anastasianadelshtehel commented on app/src/main/java/com/example/playlistmaker/presentation/playlistInfo/PlaylistInfoViewModel.kt:
```diff
@@ -0,0 +1,65 @@
+            }
+
+        return combine(playlistFlow, tracksFlow) { pl, tracks ->
+            val minutes = tracks.sumOf { it.trackTimeMillis } / 60_000L
```
При каждом обновлении playlistFlow или tracksFlow будет заново вычисляться sumOf. Для больших плейлистов это может привести к проблемам. Лучше вынести вычисление длительности в map для tracksFlow.

anastasianadelshtehel commented on app/src/main/java/com/example/playlistmaker/ui/playlistInfo/PlaylistInfoFragment.kt:
```diff
@@ -0,0 +1,422 @@
+    private var lastUi: PlaylistInfoViewModel.Ui? = null
+
+    // формат mm:ss
+    private fun formatDuration(ms: Long): String {
```
Метод можно вынести в утилитарный класс

anastasianadelshtehel commented on app/src/main/java/com/example/playlistmaker/ui/playlistInfo/PlaylistInfoFragment.kt:
```diff
@@ -0,0 +1,422 @@
+
+
+class PlaylistInfoFragment : Fragment(R.layout.fragment_playlist_info), OnTrackClickListener {
+
```
В классе много логики, можно вынести логику форматирования и построения текста в отдельные классы-утилиты или UseCase

anastasianadelshtehel commented on app/src/main/java/com/example/playlistmaker/ui/playlistInfo/PlaylistInfoFragment.kt:
```diff
@@ -0,0 +1,422 @@
+    private val args: PlaylistInfoFragmentArgs by navArgs()
+    private val viewModel: PlaylistInfoViewModel by viewModel()
+
+    private fun lerp(a: Float, b: Float, t: Float) = a + (b - a) * t.coerceIn(0f, 1f)
```
функция уже есть в Android

amanzhola committed:
Sprint 23(home work) Gr26

fixing after review:
1. Добавлен дежурный fail для сообщения: альбом пустой (Спасибо ревью! пропустил в ТЗ)
2. Исправлен PlaylistRepositoryImpl ->
методы addTrackToPlaylist и observeTracksByIds так как в базе хранится по другому (Спасибо ревью! даже и не видел, пропустил в ТЗ)
3. Вообще убран (val tracksCount: Int,) в data class Playlist так как размер намного удобнее и естественее (ленился refactoring на дубликат и тут идея(!), ведь можно ничего не ломая обновить)
4. вынесен расчёт минут внутрь tracksFlow
+ data class TracksBundle (Спасибо ! ценное замечание от ревью)
Теперь sumOf вызывается только при изменении tracksFlow, а не при каждом апдейте плейлиста
5. lerp вообщу убран (не смог использовать андроидовский)
6. Самое большое время 95% (пункт 6, остальные 14 пунтков 5%) исправления заняло затемнение при 2м BS(наконец-то, мешала инерция или оглядка на 1ый BS)
7. Для PlaylistInfoFragment созданы помощники в ui util:
DialogExt.kt, FormatUtils.kt, ImageModelUtils.kt, SnackbarExt.kt, ViewExt.kt (Опять спасибо ревью! особенно нужен был для Snackbar и просто не доходила, надо было еще в 22 сделать)
8. strings.xml + строка для альбом пуст для PlaylistInfoFragment
9. Constants.kt уввеличил MENU_DIM
10. fragment_playlist_info.xml + fail
11. Playlist.kt - убран val tracksCount: Int, точнее аккуратно заменен на размер(избежал рефакторинга!)
12. PlaylistDao (ничего не менял)
13. PlaylistInfoFragment - счищен за счет хелперов
14. PlaylistInfoViewModel.kt - исправлена счет суммы времени
15. PlaylistMappers.kt - отраженение из-за убрана val tracksCount: Int, один из параметров Playlist

amanzhola committed:
Sprint 23(home work) Gr26

fixing after review:
1. fragment_playlist_info.xml забыл добавить заглушку fail для табов. У них форма квадратная и не соотвествует форме мобильных которые прямоугольные и потому нужно другая логика для реализации соотвественно хмл для мобильных здесь вообще не проходит. Созданная и реализованная компромисс, требуется другая логика и дизайн для реализации экрана.

amanzhola committed:
Sprint 23(home work) Gr26

fixing after review:
1. fragment_playlist_info.xml забыл добавить заглушку fail для табов. У них форма квадратная и не соотвествует форме мобильных которые прямоугольные и потому нужно другая логика для реализации соотвественно хмл для мобильных здесь вообще не проходит. Созданная и реализованная компромисс, требуется другая логика и дизайн для реализации экрана.

amanzhola committed:
Sprint 23(home work) Gr26

Update:
1. Удалены
ThemeViewModel
ThemeManager

2. domain.api.base
ThemeInteraction-cleaning

3. domain.impl.base
ThemeInteractionImpl -cleaning

4. CreatePlaylistViewModel
Логика dirty изменена на иминения в сравнение с оригиналом
для CreatePlaylistFragment чтобы делать предупреждение при выходе как с создания так и с редактирования

5. MoviesViewModel
Добавил «предохранитель» на изменения при возвращении с фрагментов (можно сказать баг)

6. presentation.utils
SegmentManager
Переключение через тулбар для темы (унифицирован для фрагментов и активити[исправления оставания тулбара маин фрагмента])

7. presentation.utils
ThemeLanguageHelper
убран мусор лишние функции

8. presentation.utils
ToolbarHelper
+ fun applyThemeColors() для всех тема по тулбару
+ fun applyMainBlueColors() для маин фрагмент тема по тулбару

9. MainActivity
+ navController.addOnDestinationChangedListener (при переходах обновлять тулбар)

10. SearchFragment
убран ручной Фон тулбара и // Цвет заголовка
+ в резюме (activity as? BaseActivity)?.applyToolbarThemeColors()

11. ExtraOptionFragment
убран ручной Фон тулбара и // Цвет заголовка
+ в резюме (activity as? BaseActivity)?.applyToolbarThemeColors()

12. CreatePlaylistFragment
изменен showExitDialog + для редактирования
Главное (!) исправил упущение на конвертирование для обложки при редактировании так как не могла читать coverUri = toUriOrNull(coverPathOrUri)

13. MainFragment
убран ручной Фон тулбара и // Цвет заголовка
+ в резюме (activity as? BaseActivity)?.applyToolbarThemeColors()

14. MediaLibraryFragment
убран ручной Фон тулбара и // Цвет заголовка
+ в резюме (activity as? BaseActivity)?.applyToolbarThemeColors()

15. PlaylistInfoFragment
ensureMenuInit
исправил, оказалось забыл при выходе на edit закрыть BottomSheet
чтобы при возврате кнопка menu был кликабельным

16. SettingsFragment
убран ручной Фон тулбара и // Цвет заголовка
+ в резюме (activity as? BaseActivity)?.applyToolbarThemeColors()

17. App
убран мусор
private lateinit var themeManager: ThemeManager // 😎
🎨 Инициализация ThemeManager через Repository
fun isDarkThemeEnabled() = themeManager.repository.isDarkTheme()

18. BaseActivity
+ override fun onConfigurationChanged
+ fun applyToolbarThemeColors()

19. attr.xml
+ toolbarColor
+ toolbarContentColor

20. strings
+ string resources

21. themes.xml
+ базовые цвета тулбара для всех экранов, кроме Main
+ Глобальная тема приложения = просто наследник базы
+ Оверлей для MainActivity со синим тулбаром
перенос
menuTintColor

22. themes.xml(night)
+ базовые цвета тулбара для всех экранов, кроме Main
+ Глобальная тема приложения = просто наследник базы
+ Оверлей для MainActivity со синим тулбаром

23. activity_main.xml
убран background
